### ISSUE TITLE
Move to Clojure 1.7, reader conditionals and common code base

### ISSUE THREAD

**Issue by lprefontaine** [rbrush/clara-rules#110]
Follow up from issue 53.
The source merge operation looks relatively easy.
Ryan, I would like to split the compiler code to allow reuse of some of its code in
the tools. Will create package compiler and try to find (intelligent) name(s) for code chunks.

I added an aot profile to the project so we can easily check the impact of code modifications.

**Comment by lprefontaine** [comment]
I am almost done with code mangling I think.
Meta data it seems cannot be wrapped with conditionals. That generated some duplication.
The compiler stuff has been split, I wanted to known where we have have real issues
between clojure and clojurescript.

I moved expression related fns into compiler.expressions, helpers fns in compiler.helpers, now the
compiler name space is down to more or less 400 lines.

Pretty sure I can send a pull rq tomorrow. After that I'll tackle tests...

**Comment by lprefontaine** [comment]
Existing JVM tests run fine. Will now look at ClojureScript tests

**Comment by rbrush** [comment]
Awesome! One thing I'd like to do in a follow up is to add more to the CLJS
tests. They are pretty minimal today but will check the core functionality
of the engine at least.

**Comment by lprefontaine** [comment]
That will be my main task after I get this &$@#&$&@#& clojurescript test setup working.
Trying cljs.test but struggling with phantom etc... It seems to run forever.
Not used to this test setup yet, my expectations are probably too high :)

**Comment by lprefontaine** [comment]
Hi,
a quick question, the definitions of the record Rulebase is not the same in the Clojure and ClojureScript implementations. The ClojureScript holds queries.

Did not had the time yet to read the entire code. Is this motivated by an implementation issue or some code evolution that was not replicated in ClojureScript yet ? (or vice-versa).

I would like to unify as much as possible what can be unified in rules.cljx. The Clojure definition uses schema and the ClojureScript one does not. Would be nice to have the one using schema. The queries field could be present in Clojure even if it's not used. I would put an appropriate comment about this.

I will keep a rules.cljs and a rules.clj, obviously some ClojureScript restrictions makes this mandatory. However under the hood we sould share as much as possible between the two implementations.

I should have a working ClojureScript/cljs.test implementation in the next hour.

Comments ? Suggestions ?

**Comment by lprefontaine** [comment]
After looking at the top rules.clj/rules.cljs code I am not so sure that I would keep these separated. Most of the code is identical. Looks feasible to deal with the implementation fork there.
What do you think ?

**Comment by lprefontaine** [comment]
Got cljs unit tests to work with cljs.test and phantomjs.

I will mess with github to clone Alan's repo and push my stuff over his.
I will go ahead with a separate commit to merge rules.clj/cljs to see what can come from it.

**Comment by rbrush** [comment]
Great!

As to the question of the differences in the rulebase structure: this was
an oversight on my part and there is no good reason for it. I just updated
the CLJ version (since that's what I'm currently using in production) and
failed to make the corresponding change in CLJS. The good news is this
transition will make it so that type of mistake won't happen any more.

-Ryan

**Comment by lprefontaine** [comment]
Just found out that Alan's repo was taken from 0.8.5. 
I'll reconstruct the changes from there to bring it at par with master...
Should have checked earlier.

**Comment by lprefontaine** [comment]
Got it done. Now the commits are consistent with your master.
https://github.com/lprefontaine/clara-rules. All the existing tests are running.
I will now look at unifying rules.clj/cljs.
And see if I can solve this typedef mismatch myself :)

**Comment by lprefontaine** [comment]
Before doing anything with rules.clj/.cljs, I am slicing the engine:
helpers, impl (protocols, some defrecords), nodes (rete nodes, splitting some nodes (accumulators, transports, sessions) from the main node types, wme (for all the rest of smaller memory items).

This should prove easier to include the common pieces needs on the ClojureScript side given the current state of rules.clj and rules.cljs.

Engine.clj was above 1000 lines and that makes it harder for me to grasp (hey, I am 50+ after all : )
A bit the same kind of job as I did on the compiler.

**Comment by lprefontaine** [comment]
I pushed my last work in the above repo. I split the engine and fixed dependencies accordingly.
Tests still running fine. BTWY for cljs test the magic command is
lein cljsbuild once test

If you want a full aot check (clj and cljs):
lein with-profile aot check

I merged that duplicate defrecord thing and removed the queries argument it was not used anywhere in the cljs code.

I might work on the cljs part tomorrow but I am leaving Saturday early for Paris.
Another push is probably out of question.
I will not bring my laptop with me just the light hardware (iPhone, iPad, ...), business talks, less fun...
I am back for serious work next Wednesday morning (your night, my morning, I am in
Morocco, GMT+1).

Can you look at this and give your feedback ? I'll monitor my emails and can discuss anything you may come with.
Thanx

**Comment by rbrush** [comment]
Wow, at first glance this looks really good. There's a lot of changes but I'll find some time to dig through this tonight and tomorrow.

A good way to go through this might be for you to actually create a pull request (even if it's not quite ready for it). That way I can comment inline in the pull request on anything questions I might have regarding a segment of code, and you should get emails for those as well.

**Comment by lprefontaine** [comment]
Ok will do so while you sleep.

Ramadan started here and all restaurants/cafés are closed before sunset.

I need to stay out a bit away from the screen :)

I went out 10 mns today at most.
Time for a bowl of fresh air :)

> Wow, at first glance this looks really good. There's a lot of changes but I'll find some time to dig through this tonight and tomorrow.
> 
> A good way to go through this might be for you to actually create a pull request (even if it's not quite ready for it). That way I can comment inline in the pull request on anything questions I might have regarding a segment of code, and you should get emails for those as well. 
> 
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/rbrush/clara-rules/issues/110#issuecomment-113303611
--
Luc Prefontaine<lprefontaine@softaddicts.ca> sent by ibisMail!

**Comment by lprefontaine** [comment]
I will also find time to add comments
next week. I went for a fast and mechanical approach.


> Wow, at first glance this looks really good. There's a lot of changes but I'll find some time to dig through this tonight and tomorrow.
> 
> A good way to go through this might be for you to actually create a pull request (even if it's not quite ready for it). That way I can comment inline in the pull request on anything questions I might have regarding a segment of code, and you should get emails for those as well. 
> 
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/rbrush/clara-rules/issues/110#issuecomment-113303611
--
Luc Prefontaine<lprefontaine@softaddicts.ca> sent by ibisMail!

**Comment by lprefontaine** [comment]
Hi, I pushed the modifications in my clone according to your comments, I will now try to reorg the top level (rules.clj...) and see what level of commonality I can get here.
Let me known if you have anything else that you want me to change meanwhile.
Expect another update by tomorrow. Paris was full of tourists... a nightmare...

**Comment by lprefontaine** [comment]
Doing a first pass. I will open some of the compiler logic to the cljs side by creating or moving
some name spaces to .cljc and making the rules.cljs more aligned with the clj side.
Session creation can I think be streamlined somewhat, cljs still requiring macros but making reuse
of the low level stuff. Will test/commit by incremental steps.

**Comment by lprefontaine** [comment]
After looking at the sessions gears, I will attack it this way:
a) make rules/engine/sessions.cljc the place for common session related reusable stuff from both sides. Presently there is no such place.
b) create rules/engine/sessions/local-transport.cljc and move LocalTransport there.
c) Move and consolidate common session stuff in rules/engine/sessions.cljc
d) rework rules cljs sessions handling using common components in rules/engine/sessions.cljc
e) finally merge API in rules.cljc.
I'll deal with macros on the cljs side when I get there. Either in a separate name space or as part of the common session implementation.

After step d, I think that what will remain in rules.cljs will be almost the same as in rules.cljc.
Comments ?

**Comment by rbrush** [comment]
Seems like a good approach to me, Luc. I did spend some more time reviewing (between unpacking from last weekend's move to a new home) and I'm comfortable with merging this into master when the changes you mention are ready. We might tweak a few things after that but I don't expect anything huge.

**Comment by lprefontaine** [comment]
Fine, will add more tests on the cljs side by the end of the week.
At the end, the compilation will be clearly separated from the runtime and since the main cljs limitation is on the compilation side, it should be easier to maintain the runtime for both implementations as common parts.

I'll start next week to use this in alpha mode as an evaluation engine for my nomadic work flow engine. I'll be the first to deal with bugs :) I'll keep updating my clone as I go until we can declare this stable enough.

**Comment by lprefontaine** [comment]
Still struggling with unification of rules.clj/cljs.
I managed to get part of the inspect tool to run from cljs. The beta trees are missign but I may find a way to get this to work once I can get the tests to run. That mixed evaluation context is a bit crazy :)

**Comment by rbrush** [comment]
One note: there are some features in CLJ that I had not yet brought to CLJS, which could interfere with broad testing in CLJS. The big one is support for doing joins between variables that are not equality based.  In CLJ we convert that to a separate test node but I don't think that exists in CLJS. 

This is my shortcoming, being focused on CLJ. It might be best to just get going with existing, minimal tests and add more after we merge.

Sorry if this made things difficult for you!

**Comment by lprefontaine** [comment]
Bah, it exercise my brain a bit,

I am trying to cheat here, bring most of the CLJ code to life in CLJS hoping to level the game between the two runtimes. I'll be on the move tomorrow have to go down to Casablanca for the day.
I'll probably get through this weekend. I am lacking some debug tools in the CLJS space and that kills me a bit. This is why I got the inspector to work. I can now have an almost complete dump of session created on the CLJS side. And there's something fishy :)

Will come back Saturday with an update.

**Comment by rbrush** [comment]
Awesome! I have been bit by the exact same *out* issue before...pretty
annoying.

**Comment by lprefontaine** [comment]
I have to stop, need some sleep. I bumped on trying to unify alpha-fns... bad idea.
I kept the cljs specific code in compiler.clj. The fns are not formed in the same way and I need time to figure out why and how to reconcile this. The compiler screams otherwise trying to deal with anonymous forms.
Still trying to run the cljs tests but with the inspector usable from cljs, it should be easy to spot what is going on.
Will resume work Saturday most probably. Except rules.clj/cljs all the mods have been pushed.
Comments welcomed as usual.

**Comment by rbrush** [comment]
Seriously, don't try to do everything in one pass. If you get things to a
point where the initial tests are passing we can merge and work on the CLJS
refactoring together. I've been busy with moving to a new home lately but
am finishing up and can help more

**Comment by lprefontaine** [comment]
Just trying to get to a logical conclusion here. I'll get the original cljs tests back to life and push again.
From then we can work concurrently on adding cljs tests (and maybe extra ones regarding cross-compatibility).

I'm deep and this and I think it's worth I finish up at a point were we have a clean slate.
And don't forget that clara-rules is also to me a critical dependency :)
The beast in me has been released :)

**Comment by lprefontaine** [comment]
Ok, just pushed my last round of mods.
All clj tests run, I had to comment out two failing cljs tests, test-simple-insert-data and test-simple-insert-map.

The first relies on some rules expressed as data, through my-session-data. The data may be wrong now that both forks are merged.
The second, test-simple-insert-map uses my-session-map which specified that :type should be used to find the type. That does not seem to work.
You probably have more history to rely on than me to tell if these two tests are still legit.

There are only 2 valid .cljs files remaining (the src/obsolete folder should be removed). 
Test files... Yeah !
22 are .cljc files, 18 .clj of which 9 are.. test files... The 9 others are mostly compiler or tool centric.

At this point, you may want to move main/clojurescript/clara/macros.clj to main/clojure/clara and
collapse main/java under src/main.

I need to work tomorrow on my own library using this first shot and change my mind.
That cljs evaluation context difference was kind of hard to swallow. A change of air will be welcomed.

I think we should look seriously at unifying tests instead of cloning the clj ones.
I'll be online tomorrow if you have any questions.

**Comment by rbrush** [comment]
Overall it looks good to me. There are probably a few tweaks I would make, but I think this is something we can do incrementally after merging this back in.

I'm not sure if you're ready for this, but I went ahead and pulled your master into a local branch. The code looks good and all tests work, but I ran into an exception when trying to package things up. Looks like there is some sort of duplication in the build that is pulling in at least one file more than once. I'm not sure  if you cane across this:

```
lein install
Compiling 4 source files to /Users/rb4106/clojure/clara-rules/target/classes
Compiling ClojureScript.
Compiling "target/js/clara.js" from ["src/main/clojurescript" "src/main/clojure"]...
WARNING: map->Rulebase at line 27 is being replaced at line 27 src/main/clojure/clara/rules/compiler/codegen.cljc
WARNING: map->Explanation at line 18 is being replaced at line 18 src/main/clojure/clara/tools/inspect.cljc
Successfully compiled "target/js/clara.js" in 13.2 seconds.
java.util.zip.ZipException: duplicate entry: clara/rules.cljc
 at java.util.zip.ZipOutputStream.putNextEntry (ZipOutputStream.java:215)
    java.util.jar.JarOutputStream.putNextEntry (JarOutputStream.java:109)
    sun.reflect.GeneratedMethodAccessor88.invoke (:-1)
    sun.reflect.DelegatingMethodAccessorImpl.invoke (DelegatingMethodAccessorImpl.java:43)
    java.lang.reflect.Method.invoke (Method.java:606)
    clojure.lang.Reflector.invokeMatchingMethod (Reflector.java:93)
    clojure.lang.Reflector.invokeInstanceMethod (Reflector.java:28)
    leiningen.jar/fn (jar.clj:151)
```

Once again, many thanks for your efforts on this!

**Comment by lprefontaine** [comment]
It's fine to start to merge this.
As for the build, the warnings come from the schema lib. A fix in the next release snapshot is available. It appeared after on if the recent clojurescript release.

The jar error is new to me. Will look at it in 20 mns. I hit this years ago when packaging jar files but can't remember why.

**Comment by lprefontaine** [comment]
Wow my last post makes me look as if I was drunk :)

The warnings appeared following one of the recent Cls releases.

**Comment by lprefontaine** [comment]
Pretty sure that the duplicate zip entry is a cljsbuild issue. The hook is probably called once after the standard jar step has been done. Since we ship the source code as part of the cljsbuild target, it tries to reinsert source code files (.cljs and .cljc files) on top of what has already been done by the standard task.
I have to leave for 30/40 mns. I'll try to create a more specific jar task that will avoid this issue.

**Comment by lprefontaine** [comment]
Did it before leaving, look at my last version of project.clj.
I removed the cljsbuild hook globally and added it explicitly in all profiles except in the new jar profile.
I moved the cljsbuild target to resources/js, this allows the jar step to pick up the output of the cljs build even if the cljsbuild hook is not enabled in the jar profile.
cljsbuild should check that a file to add is not already added that would avoid the problem.

**Comment by rbrush** [comment]
So that did the trick; everything works cleanly for me know.

However, I wonder if we *want* the JAR to pick up the clara.js resulting
from the CLJS build. I'm assuming that CLJS users will generally build
their own output JS files, which includes post-processed Clara code, and
wouldn't use the generate clara.js. It wouldn't matter that much, but that
JS file is by far the biggest resource in the clara-rules JAR, inflating
its size a bit for people who don't need it. Perhaps we should just
generate it for testing purposes and not put it in a place where it is
included in the end state?

**Comment by lprefontaine** [comment]
I still need to look at this closely. I think we should aim at generating a separate cljs artifact.

I have the same issue with the clojure trace tool. I want it in cljs space I am used to it but I think it should not be part of the clojurescript runtime.

I need to chat with D.Nolan about  packaging and which road parh is best.

Meanwhile I'll open an issue wirh cljsbuild for that redundant zip entry.

**Comment by lprefontaine** [comment]
The issue with cljsbuild adding duplicate entries is already opened. I just added a comment to it.
I posted a topic on the clojurescript google group about packaging cljs artifacts.
Let us wait a bit to see what comes out from this.
I will have to experiment anyway for my own usage starting tonight :)

**Comment by lprefontaine** [comment]
Got an answer from David, just ship the source in the jar. The CLJS compiler will find these and compile them to JS. He does not recommend shipping JS compiled files.
Most probably because code generation of the CLJS compiler may change.
So we need to avoid shipping the clara.js file in the jar. Since there are no .cljs files to add to the jar (we do not include test files), I just added an exclusion to avoid adding .js files in the jar.
The jar step still compiles CLJS  (a sanity check) but the js output is not added to the jar.

Use the aot profile to insure that everything compiles down to either class files or js files before cutting a release or when you need to insure that a change to a common file still compiles in both worlds. The aot profile now compiles using advanced mode.
The js file shrinks down to around 450k instead of nearly 3 megabytes.

I pushed the mods to the project file as I write this comment.

**Comment by lprefontaine** [comment]
Another jar related modification. Because we do not run the cljsbuild hook, macros.clj was not added to the jar. Since it's the only cljs source file remaining, I moved it to main/clojure/clara.
It now gets added to the jar target. Pushed as of now.

I discovered this when I tries to use the jar in my work flow project.

**Comment by rbrush** [comment]
The comments from David match my understanding, so it's good to know. An alternative to explicitly excluding .js files from the JAR might be to simply have the CLJS compiler write the generated clara.js to a different location. It had previously been written under the target folder and therefore wasn't picked up.

Moving macros.clj to main/clojure/clara makes sense to me.

**Comment by lprefontaine** [comment]
Just moved to Clojure 1.7, it's officially out :)

**Comment by lprefontaine** [comment]
Found an issue while integrating clara with my work flow lib.
The conditions were not resolved in cljs space in the dsl fn construct-condition.
That caused some symbols already resolved to classes to be picked up instead of being identified as cljs defrecord implementations.

The cljs compiler was trying to emit code for a java class when I was requiring a cljs record from another name space. Obviously, the cljs compiler does not know how to do that and was failing on the mutlimethod emit-constant.

Elsewhere in the dsl the test is made to see if we are compiling cljs. I added the same logic.
I kept the var check first then the test on cljs compilation phase. Now it works on my side.
You may want to review and see if what I did is sane. It's f... late here (4 am) or early depending how you look at it,

I'll got hit my bed...

**Comment by rbrush** [comment]
Just took a look at the change and it looks good. (And thanks again for the many improvements!)

**Comment by lprefontaine** [comment]
My own integration work starts to lift. I am near sharing sessions between the back end and the browser using durability :) Should be done late tonight. I  have a work flow defined and by the end of tomorrow should have this running both on the server and in the browser.
It's hot here... Hard to work during the day, I don't like air conditioning, hot/cold that's not very good to me.
Test wise, I need to finish this alpha integration stage before I leave Rabat for Montreal.
I can come back to writing tests by mid-july. If I hit serious problems, I will pause to write more cljs tests targeting the problematic area but so far so good.

When do you think you will create that 0.9 branch ? Let me know so I can checkout my local repo with yours and drop the one I created in github.

**Comment by rbrush** [comment]
Luc, I took your master, squashed the commits to clean up the history a bit
(there were a lot of intermediate work-in-progress steps), and pushed it to
a 0.9 branch in my repository. If you want, we can work from that branch
moving forward. I'll merge it into clara-rules master in my repository when
we feel like it's ready for others to use.

The 0.9 branch I created has everything up to and including this commit:

Author: Luc Préfontaine <lprefontaine@softaddicts.ca>
Date:   Wed Jul 1 04:16:41 2015 +0000

    Resolve conditions in cljs space, not in clj space, otherwise we may
end with defrecord seen as java classes and the cljs compiler does not know
how to emit code for classes..


The branch can be seen here:

https://github.com/rbrush/clara-rules/tree/0.9

I will probably do some updates and refactoring on that branch myself, and
we can merge in further changes you have as they arise so we stay in synch.

Will this approach work for you? We can continue on separate branches if
that works better for the time being, but if you feel like the major
refactoring effort is over then I'd like to get this merged.

**Comment by lprefontaine** [comment]
A1 for me, major refactoring is over except potentially the tests but this has no impacts on the internals.

When I am back in Montreal, I will look at a way to reuse the same unit tests. That would be a huge gain of time if we can reach that state. 

I'll checkout that branch tomorrow. That will be the base from which I will report issues if I  find some :)

A+
Luc

**Comment by kahunamoore** [comment]
I'll give it a spin as well. Great work Luc! Glad to see this get completed - sorry my contribution dropped off there - my daughter's wedding and several other issues meant I couldn't find any free time to continue working on it.

Are there any gotchas to expect while porting an existing Clara (ClojureScript based) app to the 0.9 version?

I will close the original defect. Let me know if there is anything I can do to help. I have some more examples I can add to the clara-examples project. I might look into upgrading that project to use 0.9.

Take care.

Alan

P.S. Luc - I'm curious about your workflow library. I'm building a workflow engine as well and it sounds very much like what you are doing. Without durability in Clara, the approach I was taking w.r.t. dynamically defined/compiled rules was to re-use a lot of what figwheel/sidecar does. You can contact me at kahunamoore with a domain of coopsource org if you would like to explore collaboration on these topics.

**Comment by lprefontaine** [comment]
Alan I have no vivid example of cljs rules to migrate. I start from scratch here and my main goal eith this work on Clara was to get the same set of rules to work on both sides of the fence from the same source.

The eval restriction makes life a bit uneasy on the ClojureScript side but it's manageable.
My previous project with Clara on the Clojure side involved end users and I had to do a number of tricks to remove most of the boilerplate code. Not clear how far I can get on that road on the clojurescript side but it's not a goal in my workflow project. The primary audience is the dev teams.

My workflow library is a bit more than just a library but durability of the  embeded eval engine is essential. I chose clara but one future goal is to allow other eval engine implementations. 

My first shot at this is likely to be alive by mid-august 'de concert' with a first real life use.

The main goal here is to lower the cost to write apps. If I succeed, I have half a dozen commercial app targets on the horizon. The market here is virgin.

I am not ready to disclose anything yet, I am in a maelstrom trying to get this up asap, creating a demo app for a customer and coping with admin crap. One clue, the thing will be called LaBalboa. 
Clockwork Orange was my first choice but rights are detained by a crazy guy that sues as soon as you spell it and he's not even suing for money.

As I said, I think that trying to unify Clara unit tests would be a great leap forward. If you do have some time to spare, this is the area that IMHO should be the prime target. That will help uncover hidden issues maybe before I hit them myself :)

**Comment by lprefontaine** [comment]
Hi,

While smoking a cuban cigar and sipping an espresso at the nearby terrace,
(between this or a hammock, guess what alternative I prefer :), I reviewed the test code.

Most of it can be unified by using macros (defquey and defrule) to hide the eval issue between clj and cljs.

Many of the queries and rules used by the tests are the same.

The DSL and compilation phase might be worth dedicated tests only run in Clojure. Presently the compilation/evaluation is kind of bundled with all the other tests.

I will sync my repo with yours and checkout the 0.9 branch and start working on this.

Any suggestions ?

Expect a pull request by next Tuesday :)

**Comment by rbrush** [comment]
Yes, I think the biggest effort is going to be to convert all of the dsl/parse-rule and dsl/parse-query structures and converting them into defrule/defquery supported in CLJs. I agree that some of the compiler and DSL steps should be separated out into namespaces run only in CLJ.

If you're willing to tackle that change then I'd be happy to accept it into the code base. One thing to consider (and I know very little about) is the status of ClojureScript-in-ClojureScript. It sounds like it's bootstrapped now, which means it *should* be possible to run the rule compiler itself in CLJS, which means dsl/parse-rule and mk-session could run in CLJS as well as CLJ. 

I have no idea whether CLJS-in-CLJS is mature enough to tackle this (I suspect it isn't ready yet), but it would free us from having to move everything to defrule/defquery and it would mean one could author rules in CLJS without CLJ running at all.

**Comment by lprefontaine** [comment]
Just watched a video from euroclojure by David and it's not fully ready yet.

I have Replete beta on my ipad which uses this.

Pretty cool. But still beta :)

Redefining the queries and rules is not a big task. I'll split the core of the tests to isolate dsl and compilation test.

Later when cljs on cljs is ready, there will be no need to change this. Using defquery/defrule vs calling the dsl has no impacts.

Assuming we add tests targeting directly the DSL and the compiler.

Once we can compile in cljs it will impact stuff under the horizon so no efforts will have been lost in restructuring the tests. 

Sounds good ?

**Comment by kahunamoore** [comment]
Yes, it's the macro support that is holding things up - I don't see that changing in the short term. Thanks Luc (and Ryan) for your hard work on this. I'm more available now so let me know how I can help. Looking at the examples repo now...

**Comment by rbrush** [comment]
Sounds good!

**Comment by lprefontaine** [comment]
While I am in this, do you have a preference for the test name space ?
Right now everything is kind of floating under clara.

clara.test ?

**Comment by rbrush** [comment]
I'd want to follow conventions used in other Clojure projects. I had seen a
lot of test namespaces mirror the namespace it is testing, but just with a
test- prefix. So a test for clara.rules.compiler would go into
clara.rules.test-compiler. If conventions are changing in other Clojure
projects we could follow suit, though.

**Comment by lprefontaine** [comment]
Ok. Will rename accordingly.

**Comment by lprefontaine** [comment]
Making progress here. However lein does not support cljc test files yet. The lein patch is done, the pull request is waiting, I am trying to figure out what will be the time line to get this out in the next lein release or as a 2.x patch. Maybe they need more tests done, looks that it was not really trivial.

Meanwhile I am getting the clojure tests to run in CCW/Eclipse and anyway this is the way to got to debug them in clj space. I will complete this today and then validate them in cljs space.

If this works, I will commit this on branch 0.9 in my fork. I will see what I can do after ward to help lein folks to get this out.

**Comment by lprefontaine** [comment]
I have implemented a work around to get common tests to run until a lein release can handle cljc test files. A bit on the same path as cljs tests. I added a clj test runner under test to run common tests by requiring their name space and running tests in them individually. This applies only for tests in cljc files. Tests in clj files run as is under lein test.

I hope to complete this work by tomorrow. I am kind of late on my agenda, admin tasks this week ate much of my shedule...

**Comment by lprefontaine** [comment]
Hi,
work in progress, you can look at the following in the 0.9 of my pull request:

src/test/common/clara/rules/ <- common tests and support files for both clj and cljs
src/test/clojure/clara/rules/test_java.clj
src/test/clojure/clara/rules/test_compiler.clj
src/test/common/clara/rules/test_rules.cljc

I moved java and compiler related tests from test_rules.cljc to clj in the two last files.
All accumulator related tests are under the common root. They run successfully both in clj and cljs.

test_rules.cljc is my work in progress this weekend, the first 300 lines are converted to use defquery/defrule/defsession instead of inlining this in test fns.

I wonder if we could split these tests along some logical frontier. It will grow around 1800 lines once I de-inline queries, rules and sessions. Sometimes I can reuse the queries, rules and sessions but there are a number of definitions I have to create for which I have to forge names. This problem inflates as the number of tests grows in a single name space.

It's not a complicated task, it's just... long :)

**Comment by lprefontaine** [comment]
I am facing some issues with the defquery/defsession macros on the cljs side.
Looks like some quoting issue when a rule defines refers to a defrecord property in a condition.
The cljs compiler screams when evaluating the rule expression in a defsession. It stumbles on
undefindped vars a soon as a rule refers to a record property.

I will resume work on this tomorrow by the end of the day. I need a fresh start to pinpoint the
problem.

**Comment by rbrush** [comment]
This sort of thing can be a pain to debug. If you can see what the expanded
macros looks like, pre-compilation, it might give a hint on any bad code
being generated. Since it's complaining of a bad var, I'd speculate the
code being expanded to test the condition isn't creating a local binding
for the field name (or destructuring the record into a local binding), so
it's looking for a var since nothing is local.

That's just speculation, though. Since writing things to stdout tends to
mess with the CLJS compilation workflow, I've gone as far as dumping the
text of the expanded macro out to a file and looking at what is generated.
I'm not sure if there is a better way to do that now. The landscape is
changing pretty fast.

**Comment by lprefontaine** [comment]
In the dev name space you will find a couple of fns to get things printed on stderr or at the console
depending on the context transparently. This what I use to get macro expansion tracability.

I have a strategy, I will use it later today to try to pinpoint the issue. You are probably right about local
bindings missing. I just need to get one case fully expanded/traced from both sides of the fence and we should have a clearer picture.

Been sucked by another one of my commercial project. And dealing with the consequences of my
ironic posts on the clojure mailing list.

Should be back to it by 15:00.

**Comment by lprefontaine** [comment]
I pushed debug versions of the macros. I think we should keep these and make them part of the api.
There's no differences in the expansion of defquery and defrule between clj and cljs.

I think I have some understanding or the problem. It has to do with the evaluation done in cljs to create the session from the network in the defsession macro.
We create the network in cljs space through the defsession macro expansion.

The way I see it, when we specify a name space to create a session, that name space has been
processed by the compiler and is fully resolved.
Here in the test file I am trying to use queries and rules defined in the same name space as the session. The compiler has not finished its job yet.

My guess is that we need to quote some values in the defsession macro expansion to prevent eager evaluation in the name space.

I have to get up early (3 am) and drive to a customer site. I will get back to this tomorrow pm.
Need to crash in my bed for a few hours.

**Comment by lprefontaine** [comment]
Hi,

added more support to debug compilation issues.

Here's an example:

(defquery-debug count-joined-winds-temp
  []
  [WindSpeed (> windspeed 10) (= ?loc location)]
  [?c <- (acc/count) from [Temperature (= ?loc location)]])

(defsession-debug count-joined-winds-temp-session [count-joined-winds-temp])
 
When running tests, this brings additional output on stderr or on the cljs console.
I will send you these files by email.

It's late so I will look at these files tomorrow. I tried to move the rulebase creation out of the macro
expansion and quote it in the expansion without success.

Closer examination of the files should bring some light on this issue.

I committed the changes so you can have a look at the machinery supporting this debug mode.

**Comment by rbrush** [comment]
Thanks, I'll take a look at these tomorrow. I'm busily trying to wrap up an
unrelated project but hope to get back to this soon.

-Ryan

**Comment by lprefontaine** [comment]
Ha ! Made some progress here. I moved the code that creates the rulebase out of the expansion
of the defsession macro. Now most of it compiles except the following.

I am facing a 'No method in multimethod 'emit-constant' for dispatch value: class java.lang.Class' but this is familiar ground (remember the previous fix I made 22 days ago ?).

I'll nail this one out and it should be fine.

**Comment by lprefontaine** [comment]
I think my changes in the defsession cljs macro are not good.
The way I see it, the facts (Temperature,....) do not get resolved in cljs space,
They end up as java classes. I am tracing the context to try to understand why these are not
getting resolved. I am wondering if the require :refer ... has any effect at all.

When the rullebase is built from the macro expansion, I end up with all these compilation errors about the bindings not defined.

More things to bang my head with. I am using the test accumulator ns and commented out everything except a single session and its related queries and rules to try to nail this combined with traces in the compiler.

**Comment by rbrush** [comment]
Working with the CLJS defsession macro could get pretty complicated. At
least prior to this round of refactoring, the CLJS defsession macro was
very different than its CLJ counterpart. It actually generated most of the
code for the Rete network inline with that macro, including the functions
that reference the facts. (In the CLJ side, the defession macro is much
simpler, since we could use eval to create expressions used in the compiled
rules at runtime.)

It's hard to speculate exactly what you're running into, but printing out
the entire expansion of the CLJS defsession might provide a hint. It should
be fully valid CLJS code referencing valid CLJS structures.

If the defsession macro expansion looks fine, then I'd take a look at the
definition of the records being used. It might be worthwhile to try to
simplify things...just a minimal CLJS test that uses defsession and CLJS
records and see if it works. (Kind of like the test that existed prior to
this refactoring effort.) From there we could apply changes incrementally
and see where things break down. Also, it could be worthwhile to compare
what the defsession expansion looks like before this refactoring effort,
and see how it differs from what is being generated now.

This is uncharted territory, so we're dealing with the pain and complexity
that comes with it. I'm *almost* wrapped up with the separate project that
has been drawing all of my attention so I can get back to Clara (which,
frankly is what I would prefer to be working on anyway.)

**Comment by lprefontaine** [comment]
Hi,

I did not had time to work on this in the last two weeks... Too many things to do and too little time.
I am about to bord on the flight back to Rabat. I will have less interrupts there, I can live like a monk overthere :)

I will resume work on this next week. I want to nail that 'uncharted territory' so we can move forward with this.

**Comment by rbrush** [comment]
An update related to this bug: a number of fixes were made to the master branch in parallel to these changes that would be pretty difficult to merge due to the level of refactoring. However, I went ahead and updated master to use Clojure 1.7 and reader conditionals, and included all of the fixes in the 0.8.9 release. 

I also updated master to use cljs.test and added a test-common.cljc, which runs Clara tests under *both* Clojure and ClojureScript for the first time! There are still some shortcomings, like working through the import or require of record types between the platforms, but we should be able to start shifting more tests to a common location used by both.

**Comment by lprefontaine** [comment]
Great, I got sucked up in a number of things and could not make my head about the evaluation context
differences yet. I need more than 24 hours a day. That's been an issue since June...
Do we need the pull rq or do we consider it as obsolete ?

**Comment by stephenbrady** [comment]
Hi, appreciate all the hard work being done here. I'm not quite sure I've seen all the places where this refactoring is happening, but I'll bring this up just in case. I'd suggest taking this opportunity to consolidate the api between clj and cljs further so that a `:require-macros [clara.macros :refer [defsession defrule defquery]]` isn't necessary for cljc/cljs usage (and the attendant reader conditionals that are necessary around those parts of the ns declaration when using clara within a .cljc file).

Prismatic Schema is an example of this where they (effectively) do a `#?(:cljs (:require-macros schema.core)`.

https://github.com/prismatic/schema/blob/master/src%2Fcljx%2Fschema%2Fcore.cljx#L88

The `defsession`, `defrule`, and `defquery` macros would need to be compile environment aware, but I see that's already happening elsewhere here: https://github.com/rbrush/clara-rules/blob/master/src%2Fmain%2Fclojure%2Fclara%2Frules%2Fcompiler.clj#L49

We have many rules that are shared between clj and cljs in .cljc files, so this change would really clean up our ns declarations. I'd submit a PR, but it seems like there's a lot of code in motion right now.

**Comment by rbrush** [comment]
Hey @stephenbrady, agreed that combining these macros is a good idea. I went ahead and logged #128 to do so.

**Comment by stephenbrady** [comment]
Great! Thanks.
